internal/envoy/route.go

```diff
@@ -41,31 +42,47 @@ func Route(match *envoy_api_v2_route.RouteMatch, action *envoy_api_v2_route.Rout
 
 // RouteMatch creates a *envoy_api_v2_route.RouteMatch for the supplied *dag.Route.
 func RouteMatch(route *dag.Route) *envoy_api_v2_route.RouteMatch {
-	match := &envoy_api_v2_route.RouteMatch{
-		Headers: headerMatcher(route.HeaderConditions),
-	}
 	switch c := route.PathCondition.(type) {
 	case *dag.RegexCondition:
-		match.PathSpecifier = &envoy_api_v2_route.RouteMatch_Regex{
-			Regex: c.Regex,
-		}
+		return RouteRegex(c.Regex, route.HeaderConditions...)
 	case *dag.PrefixCondition:
-		match.PathSpecifier = &envoy_api_v2_route.RouteMatch_Prefix{
-			Prefix: c.Prefix,
-		}
+		return RoutePrefix(c.Prefix, route.HeaderConditions...)
+	case *dag.WildcardPrefixCondition:
+		return RouteWildcardPrefix(c.Prefix, route.HeaderConditions...)
 	}
-	return match
+	return nil
 }
 
 // RouteRegex returns a regex matcher.
-func RouteRegex(regex string) *envoy_api_v2_route.RouteMatch {
+func RouteRegex(regex string, headers ...dag.HeaderCondition) *envoy_api_v2_route.RouteMatch {
 	return &envoy_api_v2_route.RouteMatch{
 		PathSpecifier: &envoy_api_v2_route.RouteMatch_Regex{
 			Regex: regex,
 		},
+		Headers: headerMatcher(headers),
 	}
 }
 
+// RouteWildcardPrefix returns a regex matcher
+func RouteWildcardPrefix(prefix string, headers ...dag.HeaderCondition) *envoy_api_v2_route.RouteMatch {
+	return &envoy_api_v2_route.RouteMatch{
+		PathSpecifier: &envoy_api_v2_route.RouteMatch_Regex{
+			Regex: sanitizeRegex(prefix),
+		},
+		Headers: headerMatcher(headers),
+	}
+}
+
+func sanitizeRegex(prefix string) string {
+	// Since the QuoteMeta will replace the user entered "*" with a "\*",
+	// it makes that character want to match when we want to actually replace
+	// with a custom regex string
+	prefix = strings.ReplaceAll(regexp.QuoteMeta(prefix), "\\*", "*")
+
+	// Replace all "*" with the match-all regex
+	return fmt.Sprintf("%s.*", strings.ReplaceAll(prefix, "*", "[0-9a-zA-Z-]*"))
```

Comment:
Also, AFAIK `[0-9a-zA-Z-]` doesn't match URL path characters (e.g. % encodings). Maybe we need `[^/`]?

Reply:
>A wildcard should match at least one right? i.e. [0-9a-zA-Z-]+

Updated to match, good point. 

> Why do you always add the .* to the end of the regex? Is this because envoy always wants the whole path to match? 

Because it's a `prefix` type, need to match anything after the string. Once we implement exact match, this regex would change.

> Also, AFAIK [0-9a-zA-Z-] doesn't match URL path characters (e.g. % encodings). Maybe we need [^/]?

You can have `%` in a path? I know you can in a query string which isn't part of this check.

Reply:
Yep, eg. http://example.com/~jpeach, or http://example.com/james%20peach/. Maybe envoy has already %-decoded before the regex, but in that case, you'd still need it to match a space character.